pkg/metrics/certificates.go

```diff
@@ -45,20 +47,25 @@ func (m *Metrics) UpdateCertificate(ctx context.Context, crt *cmapi.Certificate)
 	}
 
 	m.updateCertificateStatus(key, crt)
-	m.updateCertificateExpiry(ctx, key, crt)
+	m.updateCertificateExpiry(ctx, key, crt, timeNow)
 }
 
 // updateCertificateExpiry updates the expiry time of a certificate
-func (m *Metrics) updateCertificateExpiry(ctx context.Context, key string, crt *cmapi.Certificate) {
+func (m *Metrics) updateCertificateExpiry(ctx context.Context, key string, crt *cmapi.Certificate, timeNow time.Time) {
 	expiryTime := 0.0
+	expiryTimeRelative := 0.0
 
 	if crt.Status.NotAfter != nil {
 		expiryTime = float64(crt.Status.NotAfter.Unix())
+		expiryTimeRelative = float64(crt.Status.NotAfter.Unix() - timeNow.Unix())
```

Comment:
Hi @JoshVanL, thanks for responding.

I'd like to explore, confirm, and fix. Can you point to the code / design that supports your statement?

Reply:
The following query will return the same result, albeit having the same issue that the result is only correct at the time of execution (a recently created 90 day valid cert):

```
certmanager_certificate_expiration_timestamp_seconds - time()
````

![image](https://user-images.githubusercontent.com/15893072/118661477-0db18c00-b7e7-11eb-8a13-912681911844.png)

![image](https://user-images.githubusercontent.com/15893072/118661571-215cf280-b7e7-11eb-8b4f-0c8ec5af4b57.png)

Reply:
I'm confused by your response.

Are you proposing a different approach?

Why is "metric is always out of date" a problem?

Reply:
Thanks @kit837. I think going forward we should instead expose an `certmanager_clock_time_seconds`. This means that both the calculation that you are interested can be calculated in DataDog, as well as not having the described issue with how frequently that metric would be updated.

I would expect that this metric would be defined similar to below:

```go 
	clockTimeSeconds = prometheus.NewCounterFunc(
		prometheus.CounterOpts{
			Namespace: namespace,
			Name:      "clock_time_seconds",
			Help:      "The clock time given in seconds (from 1970/01/01 UTC).",
		},
		func() float64 {
			return float64(time.Now().Unix())
		},
	)

```

This ensures that the time given is correct according to the cert-manager clock at scrape time.

Reply:
Just to clarify on my previous point about time being incorporated within a calculated metric on certificates. The `UpdateCertificate` func is only called on state changes or resyncs of Certificate resources. This means that the time reported by the metric will nearly always be offset. The suggestion above instead calls `time.Now()` _when_ a scrape occurs, and will only be offset by the scrape period, rather than the scrap period + Certificate sync period.